### PR TITLE
problem with delay

### PR DESCRIPTION
**Pull request by vanyachyzh**
https://vanyachyzh.github.io/react_todo-app-add-and-delete/

В коміті "used Promise.all" додав async await в App; поправив короткі найменування на зрозумілі; додав тип об'єкту, який повертається в postTodo; в фції removeCompletedTodos застосував Promise.all; додав кейс Filter.All в filterTodos. Сервер не працює((

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,24 +1,73 @@
...
+  type: ErrorType.None,
+};
+
+const USER_ID = 6657;
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [filter, setFilter] = useState<Filter>(Filter.All);
+  const [error, setError] = useState<Error>(initialError);
```
Comment (OlEzhA0) [inline src/App.tsx:25]:
```suggestion
  const [filter, setFilter] = useState(Filter.All);
  const [error, setError] = useState(initialError);
```

File: src/App.tsx [hunk 2]
```diff
@@ -1,24 +1,73 @@
...
 import { UserWarning } from './UserWarning';
+import { Header } from './components/Header';
+import { TodoList } from './components/TodoList';
+import { Footer } from './components/Footer';
+import { TodoNotification } from './components/TodoNotification';
+import { getTodos, filterTodos } from './api/todos';
+import { Todo } from './types/Todo';
+import { Filter } from './types/Filter';
+import { ErrorType } from './types/ErrorType';
+import { Error } from './types/Error';
```
Comment (OlEzhA0) [inline src/App.tsx:12]: you can add file `index.ts`
and add exports like

```
export * from './Todo'
export * from './Filter'


usage:

import { Todo, Filter } from './types'
```

File: src/App.tsx [hunk 3]
```diff
@@ -1,24 +1,73 @@
...
+const USER_ID = 6657;
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [filter, setFilter] = useState<Filter>(Filter.All);
+  const [error, setError] = useState<Error>(initialError);
+
+  useEffect(() => {
+    getTodos(USER_ID)
```
Comment (OlEzhA0) [inline src/App.tsx:28]: would be better to use `async/await` syntax

File: src/App.tsx [hunk 4]
```diff
@@ -1,24 +1,73 @@
...
+          ? (
+            <>
+              <TodoList
+                todos={filterTodos(todos, filter)}
+                tempTodo={tempTodo}
+                setError={setError}
+              />
+              <Footer todos={todos} setFilter={setFilter} filter={filter} />
+            </>
+          ) : null}
```
Comment (OlEzhA0) [inline src/App.tsx:62]:
```suggestion
        {todos.length > 0 && (
            <>
              <TodoList
                todos={filterTodos(todos, filter)}
                tempTodo={tempTodo}
                setError={setError}
              />
              <Footer todos={todos} setFilter={setFilter} filter={filter} />
            </>
          )}
```

File: src/api/todos.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+import { client } from '../utils/fetchClient';
+import { Filter } from '../types/Filter';
+
+export const USER_ID = 6657;
+
+export const getTodos = (userId: number) => {
+  return client.get<Todo[]>(`/todos?userId=${userId}`);
+};
+
+export const postTodo = (userId: number, todo: {}) => {
```
Comment (OlEzhA0) [inline src/api/todos.ts:11]: add type for todo

File: src/api/todos.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  return client.post(`/todos?userId=${userId}`, todo);
+};
+
+export const deleteTodo = (userId: number, todoId: number) => {
+  return client.delete(`/todos/${todoId}`);
+
+  // eslint-disable-next-line no-console
+  console.log(userId);
```
Comment (OlEzhA0) [inline src/api/todos.ts:19]: why do we need this console

File: src/api/todos.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+};
+
+export const filterTodos = (
```
Comment (OlEzhA0) [inline src/api/todos.ts:22]: would be better to ass useMemo

File: src/api/todos.ts [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+      }
+
+      return todo.completed;
+    })
+    : todos;
+
+  return visibleTodos;
+};
+
+export const countActiveTodos = (todos: Todo[]): number => {
```
Comment (OlEzhA0) [inline src/api/todos.ts:39]:
```suggestion
export const activeTodosLength = (todos: Todo[]): number => {
```

File: src/api/todos.ts [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+  const activeTodos = todos.filter(todo => !todo.completed);
+
+  return activeTodos.length;
```
Comment (OlEzhA0) [inline src/api/todos.ts:42]:
```suggestion
  return todos.filter(todo => !todo.completed).length;
```

File: src/api/todos.ts [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+};
+
+export const checkCompletedTodos = (todos: Todo[]): boolean => {
+  return todos.some(todo => todo.completed);
+};
+
+export const deleteCompletedTodos = (todos: Todo[]): void => {
+  for (let i = 0; i < todos.length; i += 1) {
```
Comment (OlEzhA0) [inline src/api/todos.ts:50]:
```suggestion
  for (let i = 0; i < todos.length; i++) {
```

File: src/components/Footer/Footer.tsx [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+  setFilter: (filterType: Filter) => void,
+};
+
+export const Footer: React.FC<Props> = ({ todos, setFilter, filter }) => {
+  const isAnyTodoCompleted = checkCompletedTodos(todos);
+  const activeTodosAmount = countActiveTodos(todos);
+
+  const changeFilter = (event: React.MouseEvent<HTMLAnchorElement>) => {
+    const filterType
+      = event.currentTarget.textContent as 'All' | 'Active' | 'Completed';
```
Comment (OlEzhA0) [inline src/components/Footer/Footer.tsx:23]: move `'All' | 'Active' | 'Completed'` to types

File: src/components/Footer/Footer.tsx [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+  };
+
+  return (
+    <footer className="todoapp__footer">
+      <span className="todo-count">
+        {`${activeTodosAmount} items left`}
+      </span>
+
+      <nav className="filter">
+        {['All', 'Active', 'Completed'].map(link => (
```
Comment (OlEzhA0) [inline src/components/Footer/Footer.tsx:37]: move this to constant

File: src/components/Header/Header.tsx [hunk 13]
```diff
@@ -0,0 +1,70 @@
+/* eslint-disable jsx-a11y/control-has-associated-label */
+import React, { useState } from 'react';
+import { postTodo, USER_ID } from '../../api/todos';
+import { Todo } from '../../types/Todo';
+import { Error } from '../../types/Error';
+import { ErrorType } from '../../types/ErrorType';
+
+type Props = {
+  setTempTodo: (value: Todo) => void,
+  setError: (v: Error) => void,
```
Comment (OlEzhA0) [inline src/components/Header/Header.tsx:10]:
```suggestion
  setError: (err: Error) => void,
```

File: src/components/Header/Header.tsx [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+      setError({
+        state: true,
+        type: ErrorType.EmptyTitle,
+      });
+
+      return;
+    }
+
+    setTempTodo({
+      id: 0,
```
Comment (OlEzhA0) [inline src/components/Header/Header.tsx:39]: why do we need to keep id for tempTodo?

File: src/api/todos.ts (hunk 10 above)
Comment (vanyachyzh) [reply src/api/todos.ts:50]: I have a linter warning **Unary operator '++' used.**

File: src/components/Header/Header.tsx (hunk 14 above)
Comment (vanyachyzh) [reply src/components/Header/Header.tsx:39]: there is such advice in readme 
image

File: src/api/todos.ts (hunk 5 above)
Comment (IvanFesenko) [reply src/api/todos.ts:12]: Not fixed. I guess it should has Todo type
```suggestion
export const postTodo = (userId: number, todo: Todo) => {
```

**Review by IvanFesenko — CHANGES_REQUESTED** [review]
Found issue. Check video
https://user-images.githubusercontent.com/61904995/225908673-405d9169-8bca-4191-a83d-09d5a17e39fc.mp4

**Comment by vanyachyzh** [comment]
I send an object without a property ID to the server so I can't use type Todo

File: src/App.tsx [hunk 15]
```diff
@@ -1,24 +1,146 @@
...
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [filterType, setFilterType] = useState(Filter.All);
+  const [error, setError] = useState(initialError);
+  const [removedTodoId, setRemovedTodoId] = useState<number | null>(null);
+
+  useEffect(() => {
+    getTodos(USER_ID)
+      .then(response => {
+        setTodos(response);
+      })
```
Comment (etojeDenys) [inline src/App.tsx:30]:
```suggestion
      .then(setTodos)
```

File: src/App.tsx [hunk 16]
```diff
@@ -1,24 +1,146 @@
...
+  };
+
+  const removeCompletedTodos = () => {
+    const completed = todos.filter(todo => todo.completed);
+    const uncompleted = todos.filter(todo => !todo.completed);
+
+    completed.forEach(todo => {
+      deleteTodo(todo.id)
+        .then(() => {
+          setTodos(uncompleted);
```
Comment (etojeDenys) [inline src/App.tsx:95]:
```suggestion
        .then(() => 
          setTodos(uncompleted)
```

File: src/api/todos.ts (hunk 5 above)
Comment (etojeDenys) [reply src/api/todos.ts:12]: still not fixed

File: src/api/todos.ts (hunk 7 above)
Comment (etojeDenys) [reply src/api/todos.ts:20]: .

File: src/api/todos.ts [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+) => {
+  const visibleTodos = [...todos];
+
+  switch (filterParam) {
+    case Filter.Active:
+      return visibleTodos.filter(todo => !todo.completed);
+    case Filter.Completed:
+      return visibleTodos.filter(todo => todo.completed);
+    default:
+      return visibleTodos;
```
Comment (etojeDenys) [inline src/api/todos.ts:32]: better to add Filter.All case, so then we can quickly understand and increase our Filter enum

**Review by etojeDenys — CHANGES_REQUESTED** [review]
looks good, few improvements

File: src/api/todos.ts (hunk 7 above)
Comment (YegorKorenkov) [reply src/api/todos.ts:20]: not fixed

File: src/api/todos.ts (hunk 17 above)
Comment (YegorKorenkov) [reply src/api/todos.ts:33]: not fixed

**Review by YegorKorenkov — CHANGES_REQUESTED** [review]

**Review by VitaliyBondarenko1982 — APPROVED** [review]

**Pull request closed without merging**